flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/TimeBoundedStreamJoinOperator.java

```diff
@@ -0,0 +1,398 @@
+		boolean upperBoundInclusive,
+		TypeSerializer<T1> leftTypeSerializer,
+		TypeSerializer<T2> rightTypeSerializer,
+		JoinedProcessFunction<T1, T2, OUT> udf
+	) {
+
+		super(udf);
+
+		this.lowerBound = lowerBound;
+		this.upperBound = upperBound;
+
+		this.inverseLowerBound = -1 * upperBound;
+		this.inverseUpperBound = -1 * lowerBound;
+
+		this.lowerBoundInclusive = lowerBoundInclusive;
+		this.upperBoundInclusive = upperBoundInclusive;
+		this.leftTypeSerializer = leftTypeSerializer;
+		this.rightTypeSerializer = rightTypeSerializer;
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+		collector = new TimestampedCollector<>(output);
+		context = new ContextImpl(userFunction);
+
+		Class<Tuple3<T1, Long, Boolean>> leftTypedTuple =
+			(Class<Tuple3<T1, Long, Boolean>>) (Class<?>) Tuple3.class;
+
+		TupleSerializer<Tuple3<T1, Long, Boolean>> leftTupleSerializer = new TupleSerializer<>(
+			leftTypedTuple,
+			new TypeSerializer[]{
+				leftTypeSerializer,
+				LongSerializer.INSTANCE,
+				BooleanSerializer.INSTANCE
+			}
+		);
+
+		Class<Tuple3<T2, Long, Boolean>> rightTypedTuple =
+			(Class<Tuple3<T2, Long, Boolean>>) (Class<?>) Tuple3.class;
+
+		TupleSerializer<Tuple3<T2, Long, Boolean>> rightTupleSerializer = new TupleSerializer<>(
+			rightTypedTuple,
+			new TypeSerializer[]{
+				rightTypeSerializer,
+				LongSerializer.INSTANCE,
+				BooleanSerializer.INSTANCE
+			}
+		);
+
+		this.leftBuffer = getRuntimeContext().getMapState(new MapStateDescriptor<>(
+			LEFT_BUFFER,
+			LongSerializer.INSTANCE,
+			new ListSerializer<>(leftTupleSerializer)
+		));
+
+		this.rightBuffer = getRuntimeContext().getMapState(new MapStateDescriptor<>(
+			RIGHT_BUFFER,
+			LongSerializer.INSTANCE,
+			new ListSerializer<>(rightTupleSerializer)
+		));
+
+		this.lastCleanupRightBuffer = getRuntimeContext().getState(new ValueStateDescriptor<>(
+			LAST_CLEANUP_RIGHT,
+			LONG_TYPE_INFO
+		));
+
+		this.lastCleanupLeftBuffer = getRuntimeContext().getState(new ValueStateDescriptor<>(
+			LAST_CLEANUP_LEFT,
+			LONG_TYPE_INFO
+		));
+	}
+
+	/**
+	 * Process a {@link StreamRecord} from the left stream. Whenever an {@link StreamRecord}
+	 * arrives at the left stream, it will get added to the left buffer. Possible join candidates
+	 * for that element will be looked up from the right buffer and if the pair lies within the
+	 * user defined boundaries, it gets collected.
+	 *
+	 * @param record An incoming record to be joined
+	 * @throws Exception Can throw an Exception during state access
+	 */
+	@Override
+	public void processElement1(StreamRecord<T1> record) throws Exception {
+
+		long leftTs = record.getTimestamp();
+		T1 leftValue = record.getValue();
+
+		addToLeftBuffer(leftValue, leftTs);
+
+		long min = leftTs + lowerBound;
+		long max = leftTs + upperBound;
+
+		// TODO: Adapt to different bucket sizes here
+		// Go over all buckets that are within the time bounds
+		for (long i = min; i <= max; i++) {
```

Comment:
I think iterate the mapState is better, data may be very sparse. What do you think?

Reply:
I was thinking we partition the MapState in buckets of a higher granularity (e.g. one minute / one hour) and additionally even expose this option as configuration to the user?

Reply:
Well, there are multiple things to consider there:

- the value of a map state entry needs to be completely de/serialized when adding a record to the list. So larger bucket also mean more CPU overhead for de/ser.
- the few buckets we have, the fewer lookups in the state are required.
- having buckets allows to skip irrelevant timestamp ranges. If there is a predicate like `left.ts between right.ts + 1.hour and right.ts + 2.hour`, there is a 1 hour range in the state that we do not have to look at. This problem will disappear once we can rely on sorted state access.
- at least the RocksDB state backend is quite efficient in iterating over the map state because it does not deserialize any data and pre-fetches the entries.

It is not obvious which approach works better in general and you can probably construct cases where either one outperforms the other. Some performance experiments would be nice for this.

Reply:
Yes but I believe that having the iteration as it is now + the configurable bucket size, then assuming that the user knows his data, he can set the granularity of the buckets so that "buckets" are not empty. This means big buckets for low throughput streams and smaller ones for "denser" streams. 

I understand that this does not solve the issue when you have skewed keys, or when the throughput for the stream fluctuates between low and high but still we allow the user to configure it.

What do you think?

Reply:
I'm not sure that the point accesses are more efficient than the entry-iteration with pre-fetching. IMO, the most promising implementation would be entry-iteration (of possibly small bucketed entries) on a sorted state to be able to exit the iterator early. Until that is possible (sorted state does not exist yet), it is pretty much guessing and use-case dependent on what approach works better.

Reply:
I agree on that.